app/assets/javascripts/discourse/app/lib/codeblock-buttons.js

```diff
@@ -0,0 +1,193 @@
+import showModal from "discourse/lib/show-modal";
+import I18n from "I18n";
+import { guidFor } from "@ember/object/internals";
+import { clipboardCopy } from "discourse/lib/utilities";
+import { iconHTML } from "discourse-common/lib/icon-library";
+
+// Use to attach copy/fullscreen buttons to a block of code, either
+// within the post stream or for a regular element that contains
+// a pre > code HTML structure.
+//
+// Usage (post):
+//
+// const cb = new CodeblockButtons({
+//   showFullscreen: true,
+//   showCopy: true,
+// });
+// cb.attachToPost(post, postElement);
+//
+// Usage (generic):
+//
+// const cb = new CodeblockButtons({
+//   showFullscreen: true,
+//   showCopy: true,
+// });
+// cb.attachToGeneric(element);
+//
+// Make sure to run .cleanup() on the instance once you are done to
+// remove click events.
+export default class CodeblockButtons {
+  constructor(opts = {}) {
+    this._codeblockButtonClickHandlers = {};
+    this._fadeCopyCodeblocksRunners = {};
+    opts = Object.assign(
+      {
+        showFullscreen: true,
+        showCopy: true,
+      },
+      opts
+    );
+
+    this.showFullscreen = opts.showFullscreen;
+    this.showCopy = opts.showCopy;
+
+    this._boundClickHandler = this._handleClick.bind(this);
+  }
+
+  attachToPost(post, postElement) {
+    let codeBlocks = this._getCodeBlocks(postElement);
+    if (!codeBlocks.length || !post) {
+      return;
+    }
+
+    this._createButtons(codeBlocks);
+    this._storeClickHandler(post.id, postElement);
+    this._addClickEvent(postElement);
+  }
+
+  attachToGeneric(element) {
+    let codeBlocks = this._getCodeBlocks(element);
+    if (!codeBlocks.length) {
+      return;
+    }
+
+    this._createButtons(codeBlocks);
+    const commandId = guidFor(element);
+    this._storeClickHandler(commandId, element);
+    this._addClickEvent(element);
+  }
+
+  cleanup() {
+    Object.values(this._codeblockButtonClickHandlers || {}).forEach((handler) =>
+      handler.removeEventListener("click", this._boundClickHandler)
+    );
+
+    Object.values(this._fadeCopyCodeblocksRunners || {}).forEach((runner) =>
+      cancel(runner)
+    );
+
+    this._codeblockButtonClickHandlers = {};
+    this._fadeCopyCodeblocksRunners = {};
+  }
+
+  _storeClickHandler(identifier, element) {
+    if (this._codeblockButtonClickHandlers[identifier]) {
+      this._codeblockButtonClickHandlers[identifier].removeEventListener(
+        "click",
+        this._boundClickHandler
+      );
+
+      delete this._codeblockButtonClickHandlers[identifier];
+    }
+
+    this._codeblockButtonClickHandlers[identifier] = element;
+  }
+
+  _getCodeBlocks(element) {
+    let codeBlocks = [];
+    try {
+      codeBlocks = element.querySelectorAll(
+        ":scope > pre > code, :scope :not(article):not(blockquote) > pre > code"
+      );
+    } catch (e) {
+      // :scope is probably not supported by this browser
+      codeBlocks = [];
+    }
```

Comment:
Looking at the feature support matrix, I don't think we need to worry about this?

Reply:
Yah I guess this was added 2 years ago, so things have probably moved on since then